### DISCUSSION TITLE
Run the command from the system terminal (Debugger)

### DISCUSSION THREAD

**Discussion by BlackRam-oss** [microsoft/vscode-discussions#627] [Extension Development QnA] [answered]
Hello some time ago through vscode and Power Shell settings I created a tool to run a type of program.
Mainly formed by these two files:

**launch.json:**
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Ren'Py: Run",
            "type": "PowerShell",
            "request": "launch",
            "script": "bin/renpy.ps1 run",
            "cwd": "${workspaceFolder}"
        },
    "inputs": [
        {
            "id": "RenPySdk",
            "description": "Paste the path to your Ren'Py SDK folder",
            "type": "promptString",
        }
    ]
}
```

**renpy.ps1:**
```powershell
param(
    [string] $Command = "run"
)
$RenPy = Get-Content (Resolve-Path ".renpy-sdk")
$RenPy = [string]::join("", ($RenPy.Split("`n")))

if ($IsWindows -or $IsLinux -or $IsMacOS) {
    if ($IsWindows) {
        Write-Output "Renpy will run on a Windows system"
        Write-Output "Using Ren'Py SDK: $RenPy\renpy.exe"
        Start-Process -FilePath "$RenPy\renpy.exe" -ArgumentList ". $Command" -Wait
    }
    if ($IsLinux) {
        Write-Output "Renpy will run on a Linux system"
        Write-Output "Using Ren'Py SDK: $RenPy/renpy.sh"
        Start-Process -FilePath "$RenPy/renpy.sh" -ArgumentList ". $Command" -Wait
    }
    if ($IsMacOS) {
        Write-Output "Renpy will run on a MacOS system"
        Write-Output "Using Ren'Py SDK: $RenPy/renpy.sh"
        Start-Process -FilePath "$RenPy/renpy.sh" -ArgumentList ". $Command" -Wait
    }
} 
else {
    Write-Output "Renpy will run on an unknown system, may not work"
    Write-Output "Using Ren'Py SDK: $RenPy/renpy.sh"
    Start-Process -FilePath "$RenPy/renpy.sh" -ArgumentList ". $Command" -Wait
}
```

it executes command through Power Shell and show the "prints" on the terminal. 
 
(_Power Shell can be replaced with the system terminal_: Power Shell was only needed for me to determine the OS used, since vscode already handles this I can not use it and run the commands directly on the system terminal.)

Now I was thinking of creating an extension that does the same thing. (I know JavaScript, but this is the first time I created a vscode extension.)

Also for now I am not going to handle breakpoints, and trying to make it as simple as possible.

I have seen Microsoft's example. 
And I set up the following debugger:
https://github.com/DRincs-Productions/renpy-vscode/blob/main/package.json#L34

the question is how do I execute a command when I press f5?

Could you please help me

**Comment by isidorn** [comment, marked as answer]
This should just work. If I understand correctly you are using a PowerShell extension, and if something does not work I suggest to create an issue here https://github.com/PowerShell/vscode-powershell

**Reply by BlackRam-oss** [reply]
Yes it works. But I would like to create my own extension, which executes those commands instead of putting the following files into the project: renpy.ps1 and launch.json.

the reason is that I am sure others might also be interested in this type of extension

I also would not want to depend on the Power Shell extension, but directly use the system terminal

(the big problem is that I couldn't find wiki to run **terminal commands** from a vscode extension)

**Reply by isidorn** [reply]
Check out the task API please.

**Comment by BlackRam-oss** [comment]
@isidorn thanks I found the wiki (I wasn't looking with the right terms):  https://code.visualstudio.com/api/extension-guides/task-provider

----

But I have a problem, I implemented following the wiki and looking  the typescript extension.

But it gives me the following error : there are no registered activities of type 'renpy'.     
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/005230c4-c811-431e-a39c-941ecf1cca54)


what did i do wrong, i don't understand. 

the implementation is as follows:

(https://github.com/DRincs-Productions/renpy-vscode/tree/3-run-the-command-from-the-system-terminal-debugger)

```ts
export function activate(context: vscode.ExtensionContext) {
	// ...
	console.log("Ren'Py VSCode extension activated!");

	// register task provider
	import('./task/taskProvider').then(module => {
		context.subscriptions.push(module.register());
	});
        // ...
}
```




````ts
import * as vscode from 'vscode';
import { Disposable } from '../utils/dispose';

// https://code.visualstudio.com/api/extension-guides/task-provider

class RenpyTaskDefinition implements vscode.TaskDefinition {
	constructor(public task: string) {
		this.task = task;
	}
	[name: string]: any;
	type: string = 'renpy'
}

/**
 * Provides tasks for building `tsconfig.json` files in a project.
 */
class RenpyTaskProvider extends Disposable implements vscode.TaskProvider {
	public constructor() {
		super();
		this._register(vscode.workspace.onDidChangeConfiguration(this.onConfigurationChanged, this));
		this.onConfigurationChanged();
	}

	public async provideTasks(token: vscode.CancellationToken): Promise<vscode.Task[]> {
		return Promise.resolve([
			new vscode.Task(
				new RenpyTaskDefinition('run'),
				vscode.TaskScope.Workspace,
				'Run',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('compile'),
				vscode.TaskScope.Workspace,
				'Re-Compile',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('rmpersistent'),
				vscode.TaskScope.Workspace,
				'Delete Persistent',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('lint'),
				vscode.TaskScope.Workspace,
				'Lint',
				'renpy',
			),
			new vscode.Task(
				new RenpyTaskDefinition('distribute'),
				vscode.TaskScope.Workspace,
				'Distribute',
				'renpy',
			),
		]);
	}

	public async resolveTask(_task: vscode.Task): Promise<vscode.Task | undefined> {
		if (!process.env.hasOwnProperty('RenPy')) {
			vscode.window.showErrorMessage("Ren'Py SDK not found. Please set the environment variable RenPy to the path of your Ren'Py SDK.");
			return undefined;
		}

		let renpy = process.env['RenPy'];
		if (renpy === undefined) {
			vscode.window.showErrorMessage("Ren'Py SDK not found. Please set the environment variable RenPy to the path of your Ren'Py SDK.");
			return undefined;
		}

		let executable = undefined;
		if (process.platform === 'win32') {
			executable = renpy + '\\renpy.exe';
		} else if (process.platform === 'linux') {
			executable = renpy + '/renpy.sh';
		} else if (process.platform === 'darwin') {
			executable = renpy + '/renpy.sh';
		} else {
			vscode.window.showErrorMessage("Platform not supported.")
			return undefined;
		}

		const task = _task.definition.task;
		// A Renpy task consists of a task and an optional file as specified in RenpyTaskDefinition
		// Make sure that this looks like a Renpy task by checking that there is a task.
		if (task) {
			// resolveTask requires that the same definition object be used.
			const definition: RenpyTaskDefinition = <any>_task.definition;
			return new vscode.Task(
				definition,
				_task.scope ?? vscode.TaskScope.Workspace,
				definition.task,
				definition.type,
				new vscode.ProcessExecution(executable, [definition.task]), // new ShellExecution
			);
		}
		return undefined;
	}

	private onConfigurationChanged(): void {
		// const type = vscode.workspace.getConfiguration('typescript.tsc').get<AutoDetect>('autoDetect');
		// this.autoDetect = typeof type === 'undefined' ? AutoDetect.on : type;
	}
}

export function register() {
	return vscode.tasks.registerTaskProvider('renpy', new RenpyTaskProvider());
}

```





```json
      "taskDefinitions": [
        {
          "type": "renpy",
          "required": [
            "task"
          ],
          "properties": {
            "task": {
              "type": "string",
              "description": "The Renpy task to customize"
            }
          }
        }
      ]
```

**Reply by r3m0t** [reply]
Hi, set breakpoints in your code to see what code is reached. Exceptions are also logged in the Output panel(make sure you check all outputs).

If you set a breakpoint at the start and use F10/F11, it can help identify the last line of code that ran without raising an exception.

I'm not sure about that use of `import().then` either, have you considered using a standard import statement?

In the repo, activation events seem correct, is your extension being activated?

**Reply by BlackRam-oss** [reply]
I did as you suggested:

![image](https://github.com/microsoft/vscode-discussions/assets/67595890/ce5b6ea2-7e6b-49e1-8186-6cf9ba3ff75d)     
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/e2c840af-d95d-41a5-905b-14fc6dcb23f0)

activation is handled as follows:
```
  "activationEvents": [
    "workspaceContains:**/*.rpy",
    "onTaskType:renpy"
  ],
```
I've done various tests, unfortunately it still doesn't work        

this is the my first extension. i am afraid i have forgotten some trivial steps

**Reply by r3m0t** [reply]
Did you put a breakpoint on the actual call to `vscode.tasks.registerTaskProvider`?

**Reply by BlackRam-oss** [reply]
here it passes us:

![image](https://github.com/microsoft/vscode-discussions/assets/67595890/bfe0e949-6086-45cb-83bb-6e7e7b01c039)
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/432ee4cf-78b6-4a3b-a4f4-bb3234125c4b)

here never passes us (even if I open the file: tasks.json):
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/202b7f4e-3f1f-41b0-b416-0e05b1ba1ab8)
![image](https://github.com/microsoft/vscode-discussions/assets/67595890/26f56593-5d03-4eae-b632-aa4dae229b20)

**Reply by BlackRam-oss** [reply]
(I renamed  register() to registerTaskProvider())
